src/module_library/root_onedim.h

```diff
@@ -0,0 +1,515 @@
+#ifndef ROOTS_ONEDIM_H
+#define ROOTS_ONEDIM_H
+
+#include <cmath>
+#include <limits>
+#include <string>
+
+namespace root_algorithm
+{
+
+// Helper function declarations
+inline bool is_close(double x, double y, double tol, double rtol);
+inline bool is_zero(double x, double tol);
+inline bool same_signs(double x, double y);
+inline bool opposite_signs(double x, double y);
+inline bool smaller(double x, double y);
+
+
+struct graph_t {
+    double x;
+    double y;
+};
+
+enum class Flag {
+    valid,  // don't terminate
+    residual_zero,
+    delta_root_zero,
+    bracket_width_zero,
+    max_iterations,
+    invalid_bracket,
+    division_by_zero
+};
+
+bool successful_termination(Flag flag);
+std::string flag_message(Flag flag);
+
+
+template <typename T>
+struct with_flag {
+    T state;
+    Flag flag;
+
+    with_flag() {}
+    with_flag(T x) : state(x), flag(Flag::valid) {}
+    with_flag(T x, Flag f) : state(x), flag(f) {}
+    with_flag(Flag f) : flag(f) {}
+
+
+};
+
+template <typename T>
+inline bool is_valid(with_flag<T> x){ return x.flag == Flag::valid; }
+
+
+/**
+ * @class result_t
+ *
+ * @brief Result from root finding algorithm.
+ *
+ * @param root The identified root.
+ *
+ * @param [in] residual The output value (root); should equal zero if
+ * successful.
+ *
+ * @param [in] iteration The number of iterations performed.
+ *
+ * @return result_t
```

Comment:
Not sure it makes sense to have documentation for a return value since this is not a function. I don't think the `residual` and `iteration` members are input arguments, either.